sdk/tests/test_value.py

```diff
@@ -0,0 +1,94 @@
+#
+# Bentobox
+# SDK - Value Tests
+#
+
+from bento.value import *
+from bento.protos.types_pb2 import Type
+
+wrap_primitive_types = [
+    (1, Type.Primitive.INT32),
+    (-1, Type.Primitive.INT32),
+    (0xFFFFFFFF + 1, Type.Primitive.INT64),
+    ("test", Type.Primitive.STRING),
+    (0.2, Type.Primitive.FLOAT64),
+    (True, Type.Primitive.BOOL),
+    (None, Type.Primitive.INVALID),
+    ([0.2], Type.Primitive.INVALID),
+]
+
+
+def test_wrap_primitive():
+    for val, expect_type in wrap_primitive_types:
+        try:
+            proto = wrap_primitive(val)
+        except TypeError:
+            assert expect_type == Type.Primitive.INVALID
+            continue
+        # check wrapped primitive
+        dtype = proto.data_type.primitive
+        assert dtype == expect_type
+        if dtype == Type.Primitive.INT32:
+            assert proto.primitive.int_32 == val
+        elif dtype == Type.Primitive.INT64:
+            assert proto.primitive.int_64 == val
+        elif dtype == Type.Primitive.FLOAT64:
+            assert proto.primitive.float_64 == val
+        elif dtype == Type.Primitive.STRING:
+            assert proto.primitive.str_val == val
+        elif dtype == Type.Primitive.BOOL:
+            assert proto.primitive.boolean == val
+        else:
+            raise TypeError(f"Untested data type {dtype}")
+
+
+def test_wrap():
+    # wrap() should support everything wrap_primitive() can wrap
+    wrap_val_types = wrap_primitive_types[:-1] + [
+        # multi dim numpy array
+        (
+            np.ones((2, 1, 3), dtype=np.int32),
+            Type.Array(dimensions=(2, 1, 3), element_type=Type.Primitive.INT32),
+        ),
+        # single dim list
+        (
+            ["yes", "no"],
+            Type.Array(dimensions=(2,), element_type=Type.Primitive.STRING),
+        ),
+        # multi dim list
+        (
+            [[True, False], [False, True]],
+            Type.Array(dimensions=(2, 2), element_type=Type.Primitive.BOOL),
+        ),
+        # finite generator
+        (
+            (x for x in [1.2, 3.4]),
+            Type.Array(dimensions=(2,), element_type=Type.Primitive.FLOAT64),
```

Comment:
Thinking about it more, generators probably shouldn't be a primitive, since they can produce a large number of values. It would be better to convert generators to their equivalent in C++ and run it there instead.

But I think we can leave it for now.

Reply:
Good point. We can think about that moving forward.